Sources/NIOSSL/SSLCallbacks.swift

```diff
@@ -173,6 +173,84 @@ public struct PSKClientIdentityResponse: Sendable {
     }
 }
 
+public struct NIOSSLClientExtensionValues: Hashable {
+    public var serverHostname: String?
+
+    public init(serverHostname: String?) {
+        self.serverHostname = serverHostname
+    }
+}
+
+/// A callback that can used to support multiple or dynamic TLS hosts.
+///
+/// When set, this callback will be invoked once per TLS hello. The provided `NIOSSLClientExtensionValues` will contain the
+/// host name indicated in the TLS client hello, while the provided `NIOSSLContext` will be the current
+/// server context being used for the handshake.
+///
+/// Within this callback, the user can create and return a new server `NIOSSLContext` for the given host,
+/// return the same context provided if it is sufficient to complete the handshake, or return `throw` in the event
+/// of an exception.
+///
+public typealias NIOSSLContextCallback = @Sendable (NIOSSLClientExtensionValues, NIOSSLContext) -> EventLoopPromise<NIOSSLContext>
+
+/// A struct that provides helpers for working with a NIOSSLContextCallback.
+internal struct CustomContextManager {
+    private let callback: NIOSSLContextCallback
+
+    internal private(set) var state: State
+
+    init(callback: @escaping NIOSSLContextCallback) {
+        self.callback = callback
+        self.state = .notStarted
+    }
+}
+
+extension CustomContextManager {
+    internal enum State {
+        case notStarted
+
+        case pendingResult
+
+        case complete(Result<NIOSSLContext, Error>)
+    }
+}
+
+extension CustomContextManager {
+    mutating func loadContext(values: NIOSSLClientExtensionValues, on connection: SSLConnection) {
+        // Run callback to load a new context
+        let promise = self.callback(values, connection.parentContext)
+
+        // Update our state to loading
+        self.state = .pendingResult
+
+        // Ensure we have an event loop
+        guard let eventLoop = connection.eventLoop else {
+            preconditionFailure("""
+                SSL_CTX_set_cert_cb was executed without an event loop assigned to the connection.
+                This should not be possible, please file an issue.
+            """)
+        }
+
+        // Ensure we have a parent handler to resume the connection
+        guard let parentHandler = connection.parentHandler else {
```

Comment:
We don't need to crash in this context: if we can't resume the connection because we have no parent handler, there are lots of good reasons for it.